dashboard/test/seeding/lesson_seeding_test.rb

```diff
@@ -0,0 +1,146 @@
+            "key": "Bogus Lesson Group 1",
+            "lessons": [
+              {
+                "key": "Bogus-Lesson-1",
+                "properties": {
+                  "overview": "overview 1"
+                }
+              },
+              {
+                "key": "Bogus-Lesson-2",
+                "properties": {
+                  "visible_after": "visible_after 2"
+                }
+              }
+            ]
+          },
+          {
+            "key": "Bogus Lesson Group 2",
+            "lessons": [
+              {
+                "key": "Bogus-Lesson-3",
+                "properties": {
+                  "overview": "overview 3",
+                  "visible_after": "visible_after 3"
+                }
+              },
+              {
+                "key": "Bogus-Lesson-4",
+                "properties": {
+                }
+              }
+            ]
+          }
+        ]
+      }
+    JSON
+
+    assert_equal expected.strip, LessonSeeding.serialize_lessons(script)
+  end
+
+  test 'seeding sets a new property' do
+    script = create_script_with_lessons(num_lesson_groups: 1, num_lessons_per_group: 1)
+
+    # Update a property, but don't save it.
+    lesson = first_lesson_from_script(script)
+    lesson.overview = 'my overview'
+
+    # Generate JSON by serializing from the script with unsaved changes, then seed from that
+    seed_lessons_from_json(LessonSeeding.serialize_lessons(script))
+
+    assert_equal 'my overview', first_lesson_from_script(Script.find_by(name: script.name)).overview
+  end
+
+  test 'seeding updates a property' do
+    script = create_script_with_lessons(num_lesson_groups: 1, num_lessons_per_group: 1)
+    lesson = first_lesson_from_script(script)
+    lesson.visible_after = '2050'
+    lesson.save!
+
+    # Update a property, but don't save it.
+    lesson.visible_after = '3000'
+
+    # Generate JSON by serializing from the script with unsaved changes, then seed from that
+    seed_lessons_from_json(LessonSeeding.serialize_lessons(script))
+
+    assert_equal '3000', first_lesson_from_script(Script.find_by(name: script.name)).visible_after
+  end
+
+  test 'seeding unsets a property' do
+    script = create_script_with_lessons(num_lesson_groups: 1, num_lessons_per_group: 1)
+    lesson = first_lesson_from_script(script)
+    lesson.overview = 'my overview'
+    lesson.visible_after = '2050'
+    lesson.save!
+
+    # Update a property, but don't save it.
+    lesson.properties.delete('overview')
+
+    # Generate JSON by serializing from the script with unsaved changes, then seed from that
+    json = LessonSeeding.serialize_lessons(script)
+    seed_lessons_from_json(json)
+
+    reloaded_lesson = first_lesson_from_script(Script.find_by(name: script.name))
+    assert_equal nil, reloaded_lesson.overview
+    assert_equal '2050', reloaded_lesson.visible_after
+  end
+
+  def first_lesson_from_script(script)
+    script.lesson_groups.map(&:lessons).flatten.first
```

Comment:
can this just be `script.lessons.first`? or is there a problem related to not having fixed lesson ordering yet?

Reply:
Interestingly, it breaks the tests as currently written (and as the code is currently written), because these two are not the same objects in memory:
```
[development] dashboard > s = Script.find_by(name: 'coursea-2020')
...
[development] dashboard > a = s.lessons.first
  Lesson Load (8.4ms)  SELECT  `stages`.* FROM `stages` INNER JOIN `lesson_groups` ON `stages`.`lesson_group_id` = `lesson_groups`.`id` WHERE `lesson_groups`.`script_id` = 97 ORDER BY `stages`.`absolute_position` ASC, `lesson_groups`.`position` ASC LIMIT 1
...
[development] dashboard > b = s.lesson_groups.map(&:lessons).flatten.first
  LessonGroup Load (1.0ms)  SELECT `lesson_groups`.* FROM `lesson_groups` WHERE `lesson_groups`.`script_id` = 97 ORDER BY `lesson_groups`.`position` ASC
  Lesson Load (1.4ms)  SELECT `stages`.* FROM `stages` WHERE `stages`.`lesson_group_id` = 105 ORDER BY `stages`.`absolute_position` ASC
  Lesson Load (1.3ms)  SELECT `stages`.* FROM `stages` WHERE `stages`.`lesson_group_id` = 106 ORDER BY `stages`.`absolute_position` ASC
  Lesson Load (1.4ms)  SELECT `stages`.* FROM `stages` WHERE `stages`.`lesson_group_id` = 107 ORDER BY `stages`.`absolute_position` ASC
  Lesson Load (1.4ms)  SELECT `stages`.* FROM `stages` WHERE `stages`.`lesson_group_id` = 108 ORDER BY `stages`.`absolute_position` ASC
...
[development] dashboard > a.object_id == b.object_id
=> false
```

These unit tests rely on making changes to the in-memory objects and serializing from those objects to json without `save`ing the changes, so we can verify that it was the seeding that creates the expected changes. So it actually matters which in-memory object we're modifying in this test.

One possible way to fix this seems to be to use the `Script.includes(:lesson_groups, :lessons).find_by(...)` syntax you showed me just now earlier. If I query for the script that way then run the same lines, `a.object_id == b.object_id` is true. I'll try changing the tests to use that so it's a little less fragile / counterintuitive.